src/katgpucbf/xbgpu/engine.py

```diff
@@ -259,9 +273,171 @@ def capture_enable(self, enable: bool = True) -> None:
+        # Actual implementation will require some kind of 'stream_id' arg,
+        # which currently contravenes the base function signature.
+        self.send_stream.enable_substream(enable=enable)
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
+        while True:
+            # Get item from the receiver loop.
+            # - Wait for the HtoD transfers to complete, then
+            # - Give the chunk back to the receiver for reuse.
+            rx_item = await self._rx_item_queue.get()
+            if rx_item is None:
+                break
+            await rx_item.async_wait_for_events()
+
+            tx_item = await self._tx_free_item_queue.get()
+            await tx_item.async_wait_for_events()
+            # TODO: Is it fine to update the timestamp like this?
+            tx_item.reset(rx_item.timestamp)
+
+            # Bind input buffers
+
+            # Queue GPU work
+            # - For now, just fill it with zeros
+            tx_item.buffer_device.zero(self._proc_command_queue)
+            tx_item.saturated.zero(self._proc_command_queue)
+
+            # Bind output buffers
+            tx_item.add_marker(self._proc_command_queue)
+            self._tx_item_queue.put_nowait(tx_item)
+
+            # Finish with the rx_item
+            rx_item.add_marker(self._proc_command_queue)
+            self.engine.free_rx_item(rx_item)
+        # When the stream is closed, if the sender loop is waiting for a tx item,
+        # it will never exit. Upon receiving this NoneType, the sender_loop can
+        # stop waiting and exit.
+        logger.debug("gpu_proc_loop completed")
+        self._tx_item_queue.put_nowait(None)
+
+    async def sender_loop(self) -> None:  # noqa: D102
+        # NOTE: This function passes the entire downloaded data to
+        # chunk.send, which then takes care of directing data to each beam's
+        # output destination.
+
+        # Get a populated TxQueueItem from the tx_item_queue
+        while True:
+            item = await self._tx_item_queue.get()
+            if item is None:
+                break
+            # The CPU doesn't need to wait, but the GPU does to ensure it
+            # won't start the download before computation is done.
+            item.enqueue_wait_for_events(self._download_command_queue)
+
+            # Get a free Chunk from the send_stream
+            chunk = await self.send_stream.get_free_chunk()
+
+            # Kick off any data downloads from the GPU
+            item.buffer_device.get_async(self._download_command_queue, chunk.data)
+            item.saturated.get_async(self._download_command_queue, chunk.saturated)
+
+            # Wait for the transfer
+            event = self._download_command_queue.enqueue_marker()
+            await katsdpsigproc.resource.async_wait_for_events([event])
+
+            # Set the Chunk timestamp
+            chunk.timestamp = item.timestamp
+            if self.send_stream.tx_enabled:
```

Comment:
Add a TODO here so that it's clear this still needs doing.

Also I reckon the clip-cnt sensors should be updated regardless of whether the data is being transmitted.